connector_salesforce/unit/rest_api_adapter.py

```diff
@@ -0,0 +1,345 @@
+# -*- coding: utf-8 -*-
+##############################################################################
+#
+#    Author: Nicolas Bessi
+#    Copyright 2014 Camptocamp SA
+#
+#    This program is free software: you can redistribute it and/or modify
+#    it under the terms of the GNU Affero General Public License as
+#    published by the Free Software Foundation, either version 3 of the
+#    License, or (at your option) any later version.
+#
+#    This program is distributed in the hope that it will be useful,
+#    but WITHOUT ANY WARRANTY; without even the implied warranty of
+#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#    GNU Affero General Public License for more details.
+#
+#    You should have received a copy of the GNU Affero General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+##############################################################################
+from __future__ import unicode_literals
+from contextlib import contextmanager
+from functools import wraps
+import logging
+try:
+    from simple_salesforce import Salesforce
+    from simple_salesforce import (SalesforceError,
+                                   SalesforceAuthenticationFailed,
+                                   SalesforceExpiredSession)
+except ImportError:
+    logger = logging.getLogger('ImportError')
+    logger.warning('Library simple_salesforce is not available')
+
+from openerp.addons.connector.unit.backend_adapter import BackendAdapter
+from . import exceptions as connector_exception
+from ..lib.date_convertion import convert_to_utc_datetime_with_tz
+from . exceptions import SalesforceSessionExpiredError
+
+_logger = logging.getLogger(__name__)
+
+
+def with_retry_on_expiration(fun):
+    """Context manager that will retry executed action
+    on session expiration only Oauth2 is truly supported"""
+    @wraps(fun)
+    def retry(*args, **kwargs):
+        try:
+            return fun(*args, **kwargs)
+        except (SalesforceExpiredSession, SalesforceSessionExpiredError):
+            msg = "Session expired retrying"
+            _logger.warning(msg)
+            return fun(*args, **kwargs)
+    return retry
+
+
+@contextmanager
+def error_handler(backend_record):
+    """Context manager that will map Salesforce error
+    to connector error.
+
+    :param backend_record: current backend record
+    """
+    try:
+        yield
+    except SalesforceAuthenticationFailed:
+        raise connector_exception.SalesforceSecurityError(
+            'An authentication error occur please validate your credentials '
+            'in backend'
+        )
+    except SalesforceExpiredSession:
+        if backend_record.authentication_method == 'oauth2':
+            backend_record.refresh_token()
+        raise SalesforceSessionExpiredError(
+            'Token expired and was refreshed job will be retried. '
```

Comment:
'Token has expired and was refreshed: the job will be retried. '
            'In case of direct action (no job), it must be manually restarted.'